[STACK] Disallow changing the device of a tensor via set_.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #18855 Simplify storage wrapping in TH.
* #18833 [STACK] Cache device on TensorImpl; clean up TensorImpl constructors.
* **#18832 [STACK] Disallow changing the device of a tensor via set_.**

This is necessary to cache the device on a TensorImpl.

Differential Revision: [D14766231](https://our.internmc.facebook.com/intern/diff/D14766231)

gchanan committed:
[STACK] Disallow changing the device of a tensor via set_.

This is necessary to cache the device on a TensorImpl.

dzhulgakov reviewed (approved):
I think you got your stacks messed up - there's a duplicated PR for it. But stamping as the restore one was fixed in another PR.

gchanan commented:
@dzhulgakov I did a previous ghstack but accidentally ctrl-C'ed in the middle.  Even though the PRs looked good, I was worried some internal state was messed up, so I decided to just close them and try again.

gchanan committed:
Update on "[STACK] Disallow changing the device of a tensor via set_."

[STACK] Disallow changing the device of a tensor via set_.

This is necessary to cache the device on a TensorImpl.

gh-metadata: pytorch pytorch 18832 gh/gchanan/10/head
